[Query] Check whether cap is fixed or not.

gichan-jang opened the pull request:
- Check whether the cap is fixed or not before setting the cap.
 
Signed-off-by: gichan <gichan2.jang@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*] Passed [ ]Failed [ ]Skipped

jaeyun-jung commented on gst/nnstreamer/tensor_query/tensor_query_client.c:
```diff
@@ -342,25 +342,49 @@ gst_tensor_query_client_get_property (GObject * object, guint prop_id,
+    }
+
+    out_caps = gst_caps_make_writable (out_caps);
+    gst_caps_set_simple (out_caps, "framerate", GST_TYPE_FRACTION, rate_n,
```
Q: Is it necessary to set framerate on srcpad?
src pad caps template is any. plz consider the case if src is audio or other mimetype that does not need framerate.

jaeyun-jung commented on gst/nnstreamer/tensor_query/tensor_query_client.c:
```diff
@@ -342,25 +342,49 @@ gst_tensor_query_client_get_property (GObject * object, guint prop_id,
+        rate_d, NULL);
+
+    if (gst_caps_is_fixed (out_caps)) {
+      gst_pad_set_caps (self->srcpad, out_caps);
```
ret = gst_pad_set_caps (self->srcpad, out_caps);

gichan-jang commented:
> Q: Is it necessary to set framerate on srcpad?
src pad caps template is any. plz consider the case if src is audio or other mimetype that does not need framerate.

I changed to print error messages and stop the pipeline when the tensors config is invalid. (e.g., framerate is not set.)

jaeyun-jung commented on gst/nnstreamer/tensor_common.c:
```diff
@@ -274,6 +274,8 @@ gst_tensor_info_validate (const GstTensorInfo * info)
 
   if (info->type == _NNS_END) {
+    nns_loge ("Failed to validate tensor info. type: %d. "
+        "Please specify tensor type. e.g., types=uint8,float32 ", info->type);
```
Recommend to use gst_tensor_get_type_string() for log message.
```
nns_loge ("Failed to validate tensor info. type: %s. "
        "Please specify tensor type. e.g., types=uint8,float32 ", GST_STR_NULL (gst_tensor_get_type_string (info->type));
```

jaeyun-jung commented on gst/nnstreamer/tensor_common.c:
```diff
@@ -794,11 +798,16 @@ gst_tensors_config_validate (const GstTensorsConfig * config)
 
   /* tensor stream format */
   if (config->format >= _NNS_TENSOR_FORMAT_END) {
+    nns_loge ("Failed to validate tensors config. format: %d. "
```
Also recommend to use gst_tensor_get_format_string() here.

dongju-chae commented on gst/nnstreamer/tensor_query/tensor_query_serversink.c:
```diff
@@ -227,16 +227,17 @@ gst_tensor_query_serversink_start (GstBaseSink * bsink)
   GstTensorQueryServerSink *sink = GST_TENSOR_QUERY_SERVERSINK (bsink);
   GstCaps *caps;
   gchar *caps_str = NULL;
+  gboolean ret = FALSE;
```
Well.. this variable seems not to be changed in runtime. Any merit to declare it?

gichan-jang commented:
@jaeyun-jung @dongju-chae 
The tensors config validation check at the start of query server sink was removed. Even if the cap is not defined at start, the cap is received from the connected element during the negation phase. 
So, in this PR, I removed setting framerate at the start of the server sink and changed it to check only if it is fixed before setting the cap on the query client source pad.

>Recommend to use gst_tensor_get_type_string() for log message
```
nns_loge ("Failed to validate tensor info. type: %s. "
        "Please specify tensor type. e.g., type=uint8.",
        GST_STR_NULL (gst_tensor_get_type_string (info->type));
```

Thanks. Since the validation function is not used in this patch, the changes will be uploaded to another PR.


> Well.. this variable seems not to be changed in runtime. Any merit to declare it?
Removed related changes. Thanks.

jaeyun-jung commented on gst/nnstreamer/tensor_query/tensor_query_client.c:
```diff
@@ -342,25 +342,33 @@ gst_tensor_query_client_get_property (GObject * object, guint prop_id,
+    } else {
+      nns_loge ("out-caps from tensor_query_serversink is not fixed. "
+          "Failed to update client src caps, out-caps: %s", caps_str);
+      return ret;
```
plz remove this line so that out_caps can be released.

gichan-jang commented:
> plz remove this line so that out_caps can be released.

Updated. Thanks.

dongju-chae reviewed (approved):
LGTM
